modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaService.java

```diff
@@ -86,31 +87,36 @@ private <R> CompletableFuture<R> sendToReplica(ClusterNode node, ReplicaRequest
             return (CompletableFuture<R>) replica.processRequest(req);
         }
 
-        return messagingService.invoke(node.address(), req, RPC_TIMEOUT).handle((response, throwable) -> {
+        CompletableFuture<R> res = new CompletableFuture<>();
+
+        messagingService.invoke(node.address(), req, RPC_TIMEOUT).handle((response, throwable) -> {
             if (throwable != null) {
                 if (throwable instanceof TimeoutException) {
-                    throw new ReplicationTimeoutException(req.groupId());
+                    res.completeExceptionally(new ReplicationTimeoutException(req.groupId()));
                 } else if (throwable instanceof PrimaryReplicaMissException) {
-                    throw (PrimaryReplicaMissException) throwable;
+                    res.completeExceptionally(throwable);
                 }
 
-                throw new ReplicationException(req.groupId(), throwable);
+                res.completeExceptionally(withCause(
```

Comment:
I want to have a guarantee, that the trace id is written to an exception message.

Reply:
This guarantee is provided by `withCause`
```
     * Creates a new exception, which type is defined by the provided {@code supplier}, with the specified {@code t} as a cause.
     * In the case when the provided cause {@code t} is an instance of {@link IgniteInternalException}
     * or {@link IgniteInternalCheckedException}, the original trace identifier and full error code are preserved.
     * Otherwise, a newly generated trace identifier and {@code defaultCode} are used.
 ```